pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,496 @@
+
+		/// Used to notify the runtime about changes that require special
+		/// treatment.
+		type ChangeGuard: ChangeGuard<
+			PoolId = Self::CollectionId,
+			ChangeId = Self::Hash,
+			Change = Self::RuntimeChange,
+		>;
+
+		/// Max size of a data collection
+		#[pallet::constant]
+		type MaxCollectionSize: Get<u32>;
+
+		/// Max number of collections
+		#[pallet::constant]
+		type MaxFeedersPerKey: Get<u32> + Parameter;
+
+		/// The weight information for this pallet extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type Collection<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		BoundedBTreeMap<T::OracleKey, OracleValuePair<T>, T::MaxCollectionSize>,
+		ValueQuery,
+	>;
+
+	/// Store the keys that are registed for this collection
+	/// Only keys registered in this stored can be used to create the collection
+	#[pallet::storage]
+	pub(crate) type Keys<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		Blake2_128Concat,
+		T::OracleKey,
+		KeyInfo<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AddedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		RemovedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		UpdatedFeeders {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::FeederId, T::MaxFeedersPerKey>,
+		},
+		UpdatedCollection {
+			collection_id: T::CollectionId,
+			keys_updated: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account who trigger the action is not a collection admin.
+		IsNotAdmin,
+
+		/// The key is not in the collection.
+		KeyNotInCollection,
+
+		/// The key is not registered
+		KeyNotRegistered,
+
+		/// Collection size reached
+		MaxCollectionSize,
+
+		/// Collection size reached
+		NoOracleCollectionChangeId,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Propose an update in the feeders associated to an specific key.
+		/// The collection will only be modified once
+		/// [`Pallet::apply_update_feeders`] be called.
+		#[pallet::weight(T::WeightInfo::propose_update_feeders(T::MaxFeedersPerKey::get()))]
+		#[pallet::call_index(0)]
+		pub fn propose_update_feeders(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::FeederId, T::MaxFeedersPerKey>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				T::IsAdmin::check((who, collection_id)),
+				Error::<T>::IsNotAdmin
+			);
+
+			transactional::with_transaction(|| {
+				let result = Self::update_feeders(collection_id, key, feeders.clone());
+
+				// We do not want to apply the mutation,
+				// only check if there is no error in applying it
+				TransactionOutcome::Rollback(result)
+			})?;
```

Comment:
Thanks!!

Yes, right now, looking at the code it's infallible; good point! The error comes from `mutate_and_remove_if_clean` which for other calls need to return an error. In this case I still need to handle the error to avoid `.unwrap()` or something dangerous.